src/frequenz/sdk/model/_manager.py

```diff
@@ -0,0 +1,273 @@
+        return ["Single"]
+
+    def _build_path(self, key: Any) -> str:
+        """Build the model path for the given key.
+
+        Args:
+            key: the key for which to build the model path.
+
+        Returns:
+            the model path.
+        """
+        return self.model_path
+
+
+@dataclass
+class DirectModelRepository(ModelRepository):
+    """Repository to manage models based on a direct configuration dictionary."""
+
+    model_config: dict[str, str]
+    """A dictionary mapping keys to the full paths of model files."""
+
+    def _get_allowed_keys(self) -> list[str]:
+        """Get the list of allowed keys for the model repository.
+
+        Returns:
+            A list of keys as specified in the model_config dictionary.
+        """
+        assert self.model_config, "Model configuration must be set."
+        return list(self.model_config.keys())
+
+    def _build_path(self, key: str) -> str:
+        """Build the model path for the given key, directly using the path from the model_config.
+
+        Args:
+            key: The key for which to build the model path, must be in the model_config.
+
+        Returns:
+            The full path to the model file associated with the given key.
+
+        Raises:
+            KeyError: If the given key is not found in the model_config.
+        """
+        assert self.model_config, "Model configuration must be set."
+        if key not in self.model_config:
+            raise KeyError(f"Key '{key}' not found in the model configuration.")
+        return self.model_config[key]
+
+
+@dataclass
+class _Model(Generic[T]):
+    """Represent a machine learning model."""
+
+    data: T
+    """The machine learning model."""
+
+    path: str
+    """The path to the model file."""
+
+
+class ModelManager(Generic[T]):
+    """Manage machine learning models.
+
+    The model manager acts as a central hub for retrieving models, reloading
+    models, and monitoring model paths for changes.
+    """
+
+    def __init__(self, model_repository: ModelRepository) -> None:
+        """Initialize the model manager.
+
+        Args:
+            model_repository: a model repository that defines the mapping
+                between model keys and paths.
+        """
+        self._models: dict[Any, _Model[T]] = {}
+
+        models_config = model_repository.get_models_config()
+        for key, model_path in models_config.items():
+            self._models[key] = _Model(data=self._load(model_path), path=model_path)
+
+        self._monitoring_task: asyncio.Task[None] = asyncio.create_task(
+            self._start_model_monitor()
+        )
+
+    def __getitem__(self, key: Any) -> T:
+        """Get a specific loaded model using the subscript operator.
+
+        Args:
+            key: the key to identify the model.
+
+        Returns:
+            the model instance corresponding to the key.
+        """
+        return self.get_model(key)
+
+    def get_model(self, key: Any = None) -> T:
```

Comment:
To be correct, this needs to get a bit more complicated if we make keys generic. We should use a sentinel class, like:

```py
class Unspecified:
    pass
```

```suggestion
    def get_model(self, key: U = Unspecified) -> T:
```